src/main.cc

```diff
@@ -1759,6 +1765,21 @@ masterReconfigureFinish()
     reconfiguring = 0;
 }
 
+/// Rotates master process logs; \sa mainRotate()
+static void
+masterRotate()
+{
+    if (AvoidSignalAction("log rotation", do_rotate))
+        return;
+
+    // XXX: When all SMP Squid processes use the same cache.log (default), each
+    // process, including this master process, rotates the same log file,
+    // resulting in K excessive rotations for K kid processes.
+    _db_rotate_log(); // cache.log
```

Comment:
The bug behind this XXX is at least as old as SMP support. This PR documents it and adds one more name rotation (for cache.log). Please see PR description for details.

Arguably, this added rotation is the only one that should be happening, but it would take a lot of work to actually achieve that single-rotation goal in SMP Squids -- we may have to refactor all kids to stop logging to files and start sending log records to this master process (via pipes or some such)!

I have an alternative suggestion for addressing this "too many rotations" problem: Let's drop all internal support for filename renaming during log rotation. Squid will still close/reopen log files, of course, but admins will need to use logrotate or a similar external tool to rename the logs.

Whether the harm added by this draft PR (again, see PR description for details) is acceptable depends, in part, on whether we are going to remove renaming support: If we are going to drop it, then this PR should go in (after any other problems are addressed, of course). This uncertainty is the primary reason I marked this PR as a draft.

----

Long-term, we probably want to move towards logging to stdout. Container-using folks would love that, and stdout is easy to understand and use in any production environment. Since stdout cannot support rotations, disabling that support now would make any future transition easier/smaller...

[^1]: Using stdout may still require refactoring all kids to send log records to this master process to ensure that records from different kids do not overlap. Thus, it is not necessarily a _simplification_ from code/development point of view. That move may be a good idea for other reasons. I am mentioning it here just because it may be used as one more reason to drop renaming support.

Reply:
Better to have all squid write to their stderr and a dedicated logging process managing delivery of those sockets to correct cache.log. There is an open wishlist to allow a log daemon helper to get cache.log.

Reply:
Glad we seem to agree regarding the future architecture -- stdout or stderr logging from kid processes, merged at a central collection point. Whether we will use stdout or stderr, and whether that central point is the master process or some dedicated daemon are all secondary details that should be finalized by those working on that feature in the future.